Обязательная часть
Будем осуществлять работу с непростым набором данных о состоянии здоровья лошадей, испытывающих кишечные колики.

Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.\

1. TItle: Horse Colic database

2. Source Information
   -- Creators: Mary McLeish & Matt Cecile
	  	Department of Computer Science
		University of Guelph
		Guelph, Ontario, Canada N1G 2W1
		mdmcleish@water.waterloo.edu
   -- Donor:    Will Taylor (taylor@pluto.arc.nasa.gov)
   -- Date:     8/6/89

3. Past Usage:
   -- Unknown

4. Relevant Information:

   -- 2 data files
      -- horse-colic.data: 300 training instances
      -- horse-colic.test: 68 test instances
   -- Possible class attributes: 24 (whether lesion is surgical)
     -- others include: 23, 25, 26, and 27
   -- Many Data types: (continuous, discrete, and nominal)

5. Number of Instances: 368 (300 for training, 68 for testing)

6. Number of attributes: 28

7. Attribute Information:

  1:  surgery?
          1 = Yes, it had surgery
          2 = It was treated without surgery

  2:  Age
          1 = Adult horse
          2 = Young (< 6 months)

  3:  Hospital Number
          - numeric id
          - the case number assigned to the horse
            (may not be unique if the horse is treated > 1 time)

  4:  rectal temperature
          - linear
          - in degrees celsius.
          - An elevated temp may occur due to infection.
          - temperature may be reduced when the animal is in late shock
          - normal temp is 37.8
          - this parameter will usually change as the problem progresses
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock
  5:  pulse
          - linear
          - the heart rate in beats per minute
          - is a reflection of the heart condition: 30 -40 is normal for adults
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate

  6:  respiratory rate
          - linear
          - normal rate is 8 to 10
          - usefulness is doubtful due to the great fluctuations

  7:  temperature of extremities
          - a subjective indication of peripheral circulation
          - possible values:
               1 = Normal
               2 = Warm
               3 = Cool
               4 = Cold
          - cool to cold extremities indicate possible shock
          - hot extremities should correlate with an elevated rectal temp.

  8:  peripheral pulse
          - subjective
          - possible values are:
               1 = normal
               2 = increased
               3 = reduced
               4 = absent
          - normal or increased p.p. are indicative of adequate circulation
            while reduced or absent indicate poor perfusion

  9:  mucous membranes
          - a subjective measurement of colour
          - possible values are:
               1 = normal pink
               2 = bright pink
               3 = pale pink
               4 = pale cyanotic
               5 = bright red / injected
               6 = dark cyanotic
          - 1 and 2 probably indicate a normal or slightly increased
            circulation
          - 3 may occur in early shock
          - 4 and 6 are indicative of serious circulatory compromise
          - 5 is more indicative of a septicemia

 10: capillary refill time
          - a clinical judgement. The longer the refill, the poorer the
            circulation
          - possible values
               1 = < 3 seconds
               2 = >= 3 seconds

 11: pain - a subjective judgement of the horse's pain level
          - possible values:
               1 = alert, no pain
               2 = depressed
               3 = intermittent mild pain
               4 = intermittent severe pain
               5 = continuous severe pain
          - should NOT be treated as a ordered or discrete variable!
          - In general, the more painful, the more likely it is to require
            surgery
          - prior treatment of pain may mask the pain level to some extent

 12: peristalsis
          - an indication of the activity in the horse's gut. As the gut
            becomes more distended or the horse becomes more toxic, the
            activity decreases
          - possible values:
               1 = hypermotile
               2 = normal
               3 = hypomotile
               4 = absent

 13: abdominal distension
          - An IMPORTANT parameter.
          - possible values
               1 = none
               2 = slight
               3 = moderate
               4 = severe
          - an animal with abdominal distension is likely to be painful and
            have reduced gut motility.
          - a horse with severe abdominal distension is likely to require
            surgery just tio relieve the pressure

 14: nasogastric tube
          - this refers to any gas coming out of the tube
          - possible values:
               1 = none
               2 = slight
               3 = significant
          - a large gas cap in the stomach is likely to give the horse
            discomfort

 15: nasogastric reflux
          - possible values
               1 = none
               2 = > 1 liter
               3 = < 1 liter
          - the greater amount of reflux, the more likelihood that there is
            some serious obstruction to the fluid passage from the rest of
            the intestine

 16: nasogastric reflux PH
          - linear
          - scale is from 0 to 14 with 7 being neutral
          - normal values are in the 3 to 4 range

 17: rectal examination - feces
          - possible values
               1 = normal
               2 = increased
               3 = decreased
               4 = absent
          - absent feces probably indicates an obstruction

 18: abdomen
          - possible values
               1 = normal
               2 = other
               3 = firm feces in the large intestine
               4 = distended small intestine
               5 = distended large intestine
          - 3 is probably an obstruction caused by a mechanical impaction
            and is normally treated medically
          - 4 and 5 indicate a surgical lesion

 19: packed cell volume
          - linear
          - the # of red cells by volume in the blood
          - normal range is 30 to 50. The level rises as the circulation
            becomes compromised or as the animal becomes dehydrated.

 20: total protein
          - linear
          - normal values lie in the 6-7.5 (gms/dL) range
          - the higher the value the greater the dehydration

 21: abdominocentesis appearance
          - a needle is put in the horse's abdomen and fluid is obtained from
            the abdominal cavity
          - possible values:
               1 = clear
               2 = cloudy
               3 = serosanguinous
          - normal fluid is clear while cloudy or serosanguinous indicates
            a compromised gut

 22: abdomcentesis total protein
          - linear
          - the higher the level of protein the more likely it is to have a
            compromised gut. Values are in gms/dL

 23: outcome
          - what eventually happened to the horse?
          - possible values:
               1 = lived
               2 = died
               3 = was euthanized

 24: surgical lesion?
          - retrospectively, was the problem (lesion) surgical?
          - all cases are either operated upon or autopsied so that
            this value and the lesion type are always known
          - possible values:
               1 = Yes
               2 = No

 25, 26, 27: type of lesion
          - first number is site of lesion
               1 = gastric
               2 = sm intestine
               3 = lg colon
               4 = lg colon and cecum
               5 = cecum
               6 = transverse colon
               7 = retum/descending colon
               8 = uterus
               9 = bladder
               11 = all intestinal sites
               00 = none
          - second number is type
               1 = simple
               2 = strangulation
               3 = inflammation
               4 = other
          - third number is subtype
               1 = mechanical
               2 = paralytic
               0 = n/a
          - fourth number is specific code
               1 = obturation
               2 = intrinsic
               3 = extrinsic
               4 = adynamic
               5 = volvulus/torsion
               6 = intussuption
               7 = thromboembolic
               8 = hernia
               9 = lipoma/slenic incarceration
               10 = displacement
               0 = n/a
 28: cp_data
          - is pathology data present for this case?
               1 = Yes
               2 = No
          - this variable is of no significance since pathology data
            is not included or collected for these cases

8. Missing values: 30% of the values are missing

In [72]:
import pandas as pd
import numpy as np
from pandas.api.types import CategoricalDtype

In [73]:
df = pd.read_csv('horse_data.csv', header = None)
df

0   1        2      3    4   5  6  7  8  9   ...     18    19 20    21 22  \
0    2   1   530101  38.50   66  28  3  3  ?  2  ...  45.00  8.40  ?     ?  2   
1    1   1   534817   39.2   88  20  ?  ?  4  1  ...     50    85  2     2  3   
2    2   1   530334  38.30   40  24  1  1  3  1  ...  33.00  6.70  ?     ?  1   
3    1   9  5290409  39.10  164  84  4  1  6  2  ...  48.00  7.20  3  5.30  2   
4    2   1   530255  37.30  104  35  ?  ?  6  2  ...  74.00  7.40  ?     ?  2   
..  ..  ..      ...    ...  ...  .. .. .. .. ..  ...    ...   ... ..   ... ..   
295  1   1   533886      ?  120  70  4  ?  4  2  ...     55    65  ?     ?  3   
296  2   1   527702  37.20   72  24  3  2  4  2  ...  44.00     ?  3  3.30  3   
297  1   1   529386  37.50   72  30  4  3  4  1  ...  60.00  6.80  ?     ?  2   
298  1   1   530612  36.50  100  24  3  3  3  1  ...  50.00  6.00  3  3.40  1   
299  1   1   534618   37.2   40  20  ?  ?  ?  ?  ...     36    62  1     1  3   

    23     24 25 26 27  
0    2  11300  0  0  2  
1    2   2208  0  0  2  
2    2      0  0  0  1  
3    1   2208  0  0  1  
4    2   4300  0  0  2  
..  ..    ... .. .. ..  
295  2   3205  0  0  2  
296  1   2208  0  0  1  
297  1   3205  0  0  2  
298  1   2208  0  0  1  
299  2   6112  0  0  2  

[300 rows x 28 columns]

In [74]:
#ввод названий столбцов
hourse_data = df.rename(columns={0:'surgery?',
                   1:'Age',#
                   2:'Hospital_Number',
                   3:'rectal_temperature',
                   4:'pulse',
                   5:'respiratory_rate',
                   6:'temperature_of_extremities',
                   7:'peripheral pulse',
                   8:'mucous_membranes',
                   9:'capillary_refill_time',
                   10:'pain',
                   11:'peristalsis',
                   12:'abdominal_distension',
                   13:'nasogastric_tube',
                   14:'nasogastric_reflux',
                   15:'nasogastric_reflux_PH',
                   16:'rectal_examination_feces',
                   17:'abdomen',
                   18:'packed_cell_volume',
                   19:'total_protein',
                   20:'abdominocentesis_appearance',
                   21:'abdomcentesis_total_protein',
                   22:'outcome',
                   23:'surgical_lesion?',
                   24:'type_of_lesion1',
                   25:'type_of_lesion2',
                   26:'type_of_lesion3',
                   27:'cp_data'                                              
                  }        
         )
hourse_data

surgery?  Age  Hospital_Number rectal_temperature pulse respiratory_rate  \
0          2    1           530101              38.50    66               28   
1          1    1           534817               39.2    88               20   
2          2    1           530334              38.30    40               24   
3          1    9          5290409              39.10   164               84   
4          2    1           530255              37.30   104               35   
..       ...  ...              ...                ...   ...              ...   
295        1    1           533886                  ?   120               70   
296        2    1           527702              37.20    72               24   
297        1    1           529386              37.50    72               30   
298        1    1           530612              36.50   100               24   
299        1    1           534618               37.2    40               20   

    temperature_of_extremities peripheral pulse mucous_membranes  \
0                            3                3                ?   
1                            ?                ?                4   
2                            1                1                3   
3                            4                1                6   
4                            ?                ?                6   
..                         ...              ...              ...   
295                          4                ?                4   
296                          3                2                4   
297                          4                3                4   
298                          3                3                3   
299                          ?                ?                ?   

    capillary_refill_time  ... packed_cell_volume total_protein  \
0                       2  ...              45.00          8.40   
1                       1  ...                 50            85   
2                       1  ...              33.00          6.70   
3                       2  ...              48.00          7.20   
4                       2  ...              74.00          7.40   
..                    ...  ...                ...           ...   
295                     2  ...                 55            65   
296                     2  ...              44.00             ?   
297                     1  ...              60.00          6.80   
298                     1  ...              50.00          6.00   
299                     ?  ...                 36            62   

    abdominocentesis_appearance abdomcentesis_total_protein outcome  \
0                             ?                           ?       2   
1                             2                           2       3   
2                             ?                           ?       1   
3                             3                        5.30       2   
4                             ?                           ?       2   
..                          ...                         ...     ...   
295                           ?                           ?       3   
296                           3                        3.30       3   
297                           ?                           ?       2   
298                           3                        3.40       1   
299                           1                           1       3   

    surgical_lesion? type_of_lesion1 type_of_lesion2 type_of_lesion3 cp_data  
0                  2           11300               0               0       2  
1                  2            2208               0               0       2  
2                  2               0               0               0       1  
3                  1            2208               0               0       1  
4                  2            4300               0               0       2  
..               ...             ...             ...             ...     ...  
295            

#из описания Horse Colic database понимаем, что это база данных коликов у лошадей. Колики — резкая боль в желудке, вызванная брожением пищи из-за неправильного кормления, содержания или тренинга животного. Заболевания, вызывающие колики, часто сопровождаются илеусом. Илеус – нарушение или прекращение прохождения содержимого по кишечному каналу вследствие обтурации, давления или нарушения его двигательной функции. Для анализа выберем столбцы: 1. Hospital Number (нужен для идентификации истории болезни),  2. Age (определим возрастную категорию),  3. abdominal_distension (ВАЖНЫЙ параметр, дискретный), 4. abdomen (дискретный), 5. nasogastric reflux PH (непрерывный), 6. packed cell volume (непрерывный), 7. type of lesion (первый столбец, остальные пустые, содержит диагноз), 8. outcome (дискретный, исход болезни).

In [75]:
analys_data = hourse_data[['Hospital_Number', 'Age', 'abdominal_distension',
                           'abdomen','nasogastric_reflux_PH',
                           'packed_cell_volume', 'type_of_lesion1', 'outcome']]
analys_data

Hospital_Number  Age abdominal_distension abdomen nasogastric_reflux_PH  \
0             530101    1                    4       5                     ?   
1             534817    1                    2       2                     ?   
2             530334    1                    1       1                     ?   
3            5290409    9                    4       ?                  5.00   
4             530255    1                    ?       ?                     ?   
..               ...  ...                  ...     ...                   ...   
295           533886    1                    ?       5                     ?   
296           527702    1                    3       4                     ?   
297           529386    1                    3       5                     ?   
298           530612    1                    3       4                     ?   
299           534618    1                    ?       1                     ?   

    packed_cell_volume  type_of_lesion1 outcome  
0                45.00            11300       2  
1                   50             2208       3  
2                33.00                0       1  
3                48.00             2208       2  
4                74.00             4300       2  
..                 ...              ...     ...  
295                 55             3205       3  
296              44.00             2208       3  
297              60.00             3205       2  
298              50.00             2208       1  
299                 36             6112       3  

[300 rows x 8 columns]

In [76]:
# расчет базовых метрик
# сначала заменим ? на NaN
analys_data = analys_data.replace(r'?', np.NaN)
# преобразуем типы данных: Hospital_Number, Age, abdominal_distension, abdomen,type_of_lesion1, outcome - это категории
# nasogastric_reflux_PH	packed_cell_volume - это числа.	type_of_lesion1, outcome - это категория.

In [77]:
analys_data = analys_data.astype({'Hospital_Number': 'category', 'Age': 'category',
                                  'abdominal_distension':'category', 'abdomen':'category',
                                  'nasogastric_reflux_PH':'float', 'packed_cell_volume':'float',
                                  'type_of_lesion1':'category', 'outcome':'category' 
                                 })
analys_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   Hospital_Number        300 non-null    category
 1   Age                    300 non-null    category
 2   abdominal_distension   244 non-null    category
 3   abdomen                182 non-null    category
 4   nasogastric_reflux_PH  53 non-null     float64 
 5   packed_cell_volume     271 non-null    float64 
 6   type_of_lesion1        300 non-null    category
 7   outcome                299 non-null    category
dtypes: category(6), float64(2)
memory usage: 22.6 KB


In [78]:
#основные метрики для числовых данных
analys_data.describe()

nasogastric_reflux_PH  packed_cell_volume
count              53.000000          271.000000
mean                4.707547           46.295203
std                 1.982311           10.419335
min                 1.000000           23.000000
25%                 3.000000           38.000000
50%                 5.000000           45.000000
75%                 6.500000           52.000000
max                 7.500000           75.000000

In [79]:
# выводы по числовым данным: 1. nasogastric_reflux_PH - много пропусков, но они могут быть заменены на среднее (т.к. размах меньше среднего значения)
# 2. packed_cell_volume - целесообразно заменить пропуски на среднее значение.

In [80]:
#анализ категориальных данных
analys_data['Hospital_Number'].value_counts()

5279822    2
529424     2
530693     2
530526     2
532349     2
          ..
534069     1
534053     1
534004     1
533983     1
518476     1
Name: Hospital_Number, Length: 284, dtype: int64

In [81]:
#много уникальных, пропусков нет
analys_data[analys_data['Hospital_Number'] == 'NaN'] 

Empty DataFrame
Columns: [Hospital_Number, Age, abdominal_distension, abdomen, nasogastric_reflux_PH, packed_cell_volume, type_of_lesion1, outcome]
Index: []

In [82]:
analys_data['Age'].value_counts(normalize=True)

1    0.92
9    0.08
Name: Age, dtype: float64

9 - явная ошибка, мое предположение - что 9 надо заменить на 2 (молодые лошади), пропусков нет. На основе имеющихся данных следует, что 92% это взрослые лошади, 8% молодые. 

In [83]:
analys_data['abdominal_distension'].value_counts(normalize=True)

1    0.311475
3    0.266393
2    0.266393
4    0.155738
Name: abdominal_distension, dtype: float64

данные представлены корректно, 56 пропусков. Из имеющихся данных следует, что у 31% лошадей не было вздутия живота, у 26% - незначительно, у еще 26% - умеренно, и только у 15% - сильное вздутие.

In [84]:
analys_data['abdomen'].value_counts(normalize=True, sort=False)

1    0.153846
2    0.104396
3    0.071429
4    0.236264
5    0.434066
Name: abdomen, dtype: float64

данные представлены корректно, 118 пропусков. Из имеющихся данных следует, что у 15% нормальный живот, у 10% - другое, у 7% - твердые фекалии в толстой кишке, у 23% - растянутая тонкая кишка, у 43% - раздутая толстая кишка.

In [85]:
analys_data['type_of_lesion1'].value_counts(normalize=True, sort=True).head(5)

0       0.186667
3111    0.110000
3205    0.096667
2208    0.066667
2205    0.043333
Name: type_of_lesion1, dtype: float64

данные представлены корректно, пропусков нет. Распределение диагнозов (ТОП5):
0 - нет диагноза (остутствует заболевание) - 18% (ВОЗМОЖНО ЭТО ТОЖЕ ПРОПУСК); 
3111 - поражение толстой кишки, простое, механическое, обтурация(закупорка) - 11%;
3205 - поражение толстой кишки, защемление, заворот/кручение - 9%;
2208 - кишечник, защемление, грыжа - 6%;
2205 - кишечник, защемление, заворот/кручение - 4%.

In [86]:
analys_data['outcome'].value_counts(normalize=True, sort=True)

1    0.595318
2    0.257525
3    0.147157
Name: outcome, dtype: float64

данные представлены корректно, 1 пропуск. 60% - выжили, 25% - умерли, 15% - усыплены.

Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [87]:
analys_data.describe()

nasogastric_reflux_PH  packed_cell_volume
count              53.000000          271.000000
mean                4.707547           46.295203
std                 1.982311           10.419335
min                 1.000000           23.000000
25%                 3.000000           38.000000
50%                 5.000000           45.000000
75%                 6.500000           52.000000
max                 7.500000           75.000000

In [88]:
q1 = analys_data['nasogastric_reflux_PH'].quantile(0.25)
q3 = analys_data['nasogastric_reflux_PH'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = analys_data[analys_data['nasogastric_reflux_PH'].between(lower_bound, upper_bound, inclusive=True)].sort_values('nasogastric_reflux_PH')
remove_outliers.head()

Hospital_Number Age abdominal_distension abdomen  nasogastric_reflux_PH  \
97           534145   1                    3       5                    1.0   
90           534197   1                  NaN       1                    1.0   
131          530670   1                    2     NaN                    1.5   
172          535292   1                    3     NaN                    2.0   
276          534787   1                    2     NaN                    2.0   

     packed_cell_volume type_of_lesion1 outcome  
97                 55.0            3205       3  
90                 53.0            2322       1  
131                44.0            1400       2  
172                57.0            2206       1  
276                37.0            3111       3

In [89]:
pd.concat([analys_data,remove_outliers]).drop_duplicates(keep=False).value_counts()

Series([], dtype: int64)

выбросов в столбце nasogastric_reflux_PH нет.

In [90]:
q1 = analys_data['packed_cell_volume'].quantile(0.25)
q3 = analys_data['packed_cell_volume'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = analys_data[analys_data['packed_cell_volume'].between(lower_bound, upper_bound, inclusive=True)].sort_values('packed_cell_volume')
remove_outliers.head()

Hospital_Number Age abdominal_distension abdomen  nasogastric_reflux_PH  \
190          528742   1                    1       3                    NaN   
56           528872   1                  NaN     NaN                    NaN   
65           534885   1                    2       4                    NaN   
287         5305129   9                  NaN     NaN                    NaN   
189         5299603   1                    1       4                    3.0   

     packed_cell_volume type_of_lesion1 outcome  
190                23.0            3133       1  
56                 24.0            3112       1  
65                 26.0            2208       1  
287                28.0               0       1  
189                30.0           31110       1

In [91]:
pd.concat([analys_data,remove_outliers]).drop_duplicates(keep=False).value_counts('packed_cell_volume')

packed_cell_volume
75.0    2
74.0    1
dtype: int64

In [92]:
remove_outliers.packed_cell_volume.describe()

count    268.000000
mean      45.977612
std       10.031666
min       23.000000
25%       38.000000
50%       44.500000
75%       51.250000
max       73.000000
Name: packed_cell_volume, dtype: float64

In [93]:
analys_data.packed_cell_volume.describe()

count    271.000000
mean      46.295203
std       10.419335
min       23.000000
25%       38.000000
50%       45.000000
75%       52.000000
max       75.000000
Name: packed_cell_volume, dtype: float64

Т.к. существенно базовые метрики не изменились, выбросы не удаляем.

In [94]:
analys_data

Hospital_Number Age abdominal_distension abdomen  nasogastric_reflux_PH  \
0            530101   1                    4       5                    NaN   
1            534817   1                    2       2                    NaN   
2            530334   1                    1       1                    NaN   
3           5290409   9                    4     NaN                    5.0   
4            530255   1                  NaN     NaN                    NaN   
..              ...  ..                  ...     ...                    ...   
295          533886   1                  NaN       5                    NaN   
296          527702   1                    3       4                    NaN   
297          529386   1                    3       5                    NaN   
298          530612   1                    3       4                    NaN   
299          534618   1                  NaN       1                    NaN   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

Задание 3. Работа с пропусками
Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [95]:
# для 'Age' заменяем 9 на 2, пропусков нет
fill_mean = analys_data
fill_mean

Hospital_Number Age abdominal_distension abdomen  nasogastric_reflux_PH  \
0            530101   1                    4       5                    NaN   
1            534817   1                    2       2                    NaN   
2            530334   1                    1       1                    NaN   
3           5290409   9                    4     NaN                    5.0   
4            530255   1                  NaN     NaN                    NaN   
..              ...  ..                  ...     ...                    ...   
295          533886   1                  NaN       5                    NaN   
296          527702   1                    3       4                    NaN   
297          529386   1                    3       5                    NaN   
298          530612   1                    3       4                    NaN   
299          534618   1                  NaN       1                    NaN   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [96]:
fill_mean['Age'] = analys_data.Age.replace(9,2)

In [97]:
fill_mean

Hospital_Number Age abdominal_distension abdomen  nasogastric_reflux_PH  \
0            530101   1                    4       5                    NaN   
1            534817   1                    2       2                    NaN   
2            530334   1                    1       1                    NaN   
3           5290409   2                    4     NaN                    5.0   
4            530255   1                  NaN     NaN                    NaN   
..              ...  ..                  ...     ...                    ...   
295          533886   1                  NaN       5                    NaN   
296          527702   1                    3       4                    NaN   
297          529386   1                    3       5                    NaN   
298          530612   1                    3       4                    NaN   
299          534618   1                  NaN       1                    NaN   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [98]:
# для 'abdominal_distension' заменяем пропуски на no_info

In [99]:
fill_mean['abdominal_distension'] = analys_data.abdominal_distension.astype('object').fillna('no_info').astype('category')
fill_mean

Hospital_Number Age abdominal_distension abdomen  nasogastric_reflux_PH  \
0            530101   1                    4       5                    NaN   
1            534817   1                    2       2                    NaN   
2            530334   1                    1       1                    NaN   
3           5290409   2                    4     NaN                    5.0   
4            530255   1              no_info     NaN                    NaN   
..              ...  ..                  ...     ...                    ...   
295          533886   1              no_info       5                    NaN   
296          527702   1                    3       4                    NaN   
297          529386   1                    3       5                    NaN   
298          530612   1                    3       4                    NaN   
299          534618   1              no_info       1                    NaN   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [100]:
# для abdomen также заменяем на no_info

In [101]:
fill_mean['abdomen'] = analys_data.abdomen.astype('object').fillna('no_info').astype('category')
fill_mean

Hospital_Number Age abdominal_distension  abdomen  nasogastric_reflux_PH  \
0            530101   1                    4        5                    NaN   
1            534817   1                    2        2                    NaN   
2            530334   1                    1        1                    NaN   
3           5290409   2                    4  no_info                    5.0   
4            530255   1              no_info  no_info                    NaN   
..              ...  ..                  ...      ...                    ...   
295          533886   1              no_info        5                    NaN   
296          527702   1                    3        4                    NaN   
297          529386   1                    3        5                    NaN   
298          530612   1                    3        4                    NaN   
299          534618   1              no_info        1                    NaN   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [102]:
# для nasogastric_reflux_PH заменяем на моду

In [103]:
fill_mean['nasogastric_reflux_PH'].fillna(analys_data['nasogastric_reflux_PH'].mode()[0], inplace=True)
fill_mean

Hospital_Number Age abdominal_distension  abdomen  nasogastric_reflux_PH  \
0            530101   1                    4        5                    2.0   
1            534817   1                    2        2                    2.0   
2            530334   1                    1        1                    2.0   
3           5290409   2                    4  no_info                    5.0   
4            530255   1              no_info  no_info                    2.0   
..              ...  ..                  ...      ...                    ...   
295          533886   1              no_info        5                    2.0   
296          527702   1                    3        4                    2.0   
297          529386   1                    3        5                    2.0   
298          530612   1                    3        4                    2.0   
299          534618   1              no_info        1                    2.0   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [104]:
# для packed_cell_volume заменяем на среднее

In [105]:
fill_mean['packed_cell_volume'].fillna(analys_data['packed_cell_volume'].mean(), inplace=True)
analys_data

Hospital_Number Age abdominal_distension  abdomen  nasogastric_reflux_PH  \
0            530101   1                    4        5                    2.0   
1            534817   1                    2        2                    2.0   
2            530334   1                    1        1                    2.0   
3           5290409   2                    4  no_info                    5.0   
4            530255   1              no_info  no_info                    2.0   
..              ...  ..                  ...      ...                    ...   
295          533886   1              no_info        5                    2.0   
296          527702   1                    3        4                    2.0   
297          529386   1                    3        5                    2.0   
298          530612   1                    3        4                    2.0   
299          534618   1              no_info        1                    2.0   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [106]:
#type_of_lesion1 - пропусков нет.

In [107]:
#outcome - 1 пропуск, заменяем на "no info"
fill_mean['outcome'] = analys_data.outcome.astype('object').fillna('no_info').astype('category')
fill_mean

Hospital_Number Age abdominal_distension  abdomen  nasogastric_reflux_PH  \
0            530101   1                    4        5                    2.0   
1            534817   1                    2        2                    2.0   
2            530334   1                    1        1                    2.0   
3           5290409   2                    4  no_info                    5.0   
4            530255   1              no_info  no_info                    2.0   
..              ...  ..                  ...      ...                    ...   
295          533886   1              no_info        5                    2.0   
296          527702   1                    3        4                    2.0   
297          529386   1                    3        5                    2.0   
298          530612   1                    3        4                    2.0   
299          534618   1              no_info        1                    2.0   

     packed_cell_volume type_of_lesion1 outcome  
0                  45.0           11300       2  
1                  50.0            2208       3  
2                  33.0               0       1  
3                  48.0            2208       2  
4                  74.0            4300       2  
..                  ...             ...     ...  
295                55.0            3205       3  
296                44.0            2208       3  
297                60.0            3205       2  
298                50.0            2208       1  
299                36.0            6112       3  

[300 rows x 8 columns]

In [108]:
fill_mean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype   
---  ------                 --------------  -----   
 0   Hospital_Number        300 non-null    category
 1   Age                    300 non-null    category
 2   abdominal_distension   300 non-null    category
 3   abdomen                300 non-null    category
 4   nasogastric_reflux_PH  300 non-null    float64 
 5   packed_cell_volume     300 non-null    float64 
 6   type_of_lesion1        300 non-null    category
 7   outcome                300 non-null    category
dtypes: category(6), float64(2)
memory usage: 22.7 KB


Конец!